app/src/main/java/com/asfoundation/wallet/wallet_validation/generic/CodeValidationPresenter.kt

```diff
@@ -0,0 +1,201 @@
+    )
+  }
+
+  private fun checkReferralAvailability() {
+    disposables.add(
+        Single.just(true)//TODO Change when Microservices implements the changes
+            .subscribeOn(networkScheduler)
+            .observeOn(viewScheduler)
+            .doOnSuccess { handleReferralStatus(it) }
+            .subscribe()
+    )
+  }
+
+  private fun handleReferralStatus(eligible: Boolean) {
+    if (eligible) {
+      view.showReferralEligible()
+    } else {
+      view.showReferralIneligible()
+    }
+  }
+
+  private fun handleNext(status: WalletValidationStatus,
+                         validationInfo: ValidationInfo) {
+    when (status) {
+      WalletValidationStatus.SUCCESS -> checkReferralAvailability()
+      WalletValidationStatus.INVALID_INPUT -> handleError(
+          R.string.verification_insert_code_error, validationInfo)
+      WalletValidationStatus.INVALID_PHONE ->
+        activity?.showPhoneValidationView(validationInfo.countryCode, validationInfo.phoneNumber,
+            R.string.verification_insert_code_error_common)
+      WalletValidationStatus.DOUBLE_SPENT ->
+        checkReferralAvailability()
+      WalletValidationStatus.GENERIC_ERROR -> handleError(R.string.unknown_error, validationInfo)
+      WalletValidationStatus.NO_NETWORK -> {
+        view.hideKeyboard()
+        view.showNoInternetView()
+      }
+    }
+  }
+
+  private fun handleError(errorMessage: Int,
+                          validationInfo: ValidationInfo) {
+    activity?.showCodeValidationView(validationInfo, errorMessage)
+  }
+
+  private fun handleValuesChange() {
+    disposables.add(
+        Observable.combineLatest(
+            view.getFirstChar(),
+            view.getSecondChar(),
+            view.getThirdChar(),
+            view.getFourthChar(),
+            view.getFifthChar(),
+            view.getSixthChar(),
+            Function6 { first: String, second: String, third: String, fourth: String, fifth: String, sixth: String ->
+              if (isValidInput(first, second, third, fourth, fifth, sixth)) {
+                view.setButtonState(true)
+              } else {
+                view.setButtonState(false)
+              }
+            }).subscribe { })
+  }
+
+  private fun isValidInput(first: String, second: String, third: String,
+                           fourth: String, fifth: String, sixth: String): Boolean {
+    return first.isNotBlank() &&
+        second.isNotBlank() &&
+        third.isNotBlank() &&
+        fourth.isNotBlank() &&
+        fifth.isNotBlank() &&
+        sixth.isNotBlank()
+  }
+
+  private fun handleCode() {
+    disposables.add(view.getFirstChar().filter { it.isNotBlank() }.doOnNext {
+      view.moveToNextView(1)
+    }.subscribe())
+
+    disposables.add(view.getSecondChar().filter { it.isNotBlank() }.doOnNext {
+      view.moveToNextView(2)
+    }.subscribe())
+
+    disposables.add(view.getThirdChar().filter { it.isNotBlank() }.doOnNext {
+      view.moveToNextView(3)
+    }.subscribe())
```

Comment:
CodeStyle: Since our code style is kinda broken, can you do a break on .doOnNext